lib/puma/server.rb

```diff
@@ -116,12 +119,22 @@ def uncork_socket(socket)
         rescue IOError, SystemCallError
         end
       end
+
+      def closed_socket?(socket)
+        tcp_info = socket.getsockopt(Socket::SOL_TCP, Socket::TCP_INFO)
+        state = tcp_info.unpack("C".freeze)[0]
```

Comment:
You can see the definition of the tcp info struct [here](https://github.com/torvalds/linux/blob/8fa3b6f9392bf6d90cb7b908e07bd90166639f0a/include/uapi/linux/tcp.h#L163).